File: routes/payment.js
const axios = require('axios');

class SelcomPayment {
    SELCOM_API_KEY = "SWAHILI-WsGHweDFyW5OOiAs";
    SELCOM_API_SECRET = "5e9a3546-78cd-4dfa-bb30-7de9a7eb69b5";
    SELCOM_BASE_URL = "https://apigwtest.selcommobile.com";

    constructor() {
        this.timestamp = new Date().toISOString()
    }

    getDigest(data) {
        var message = 'timestamp=' + this.timestamp;
        for (var d in data) {
            message = message + '&' + d + '=' + data[`${d}`];
        }
        var CryptoJS = require("crypto-js");
        var Base64 = require("crypto-js/enc-base64");
        var hmacString = CryptoJS.HmacSHA256(message, this.SELCOM_API_SECRET);
        var base64Hmac = Base64.stringify(hmacString);
        return base64Hmac;
    }

    getHeaders(data) {
        var fields = Object.keys(data);
        var headers = {
            "Content-type": "application/json",
            "Cache-Control": "no-cache",
            "Accept": "application/json",
            "Authorization": "SELCOM U1dBSElMSS1Xc0dId2VERnlXNU9PaUFz",
            "Digest-Method": "HS256",
            "Digest": this.getDigest(data),
            "Timestamp": this.timestamp,
            "Signed-Fields": fields.join()
        }
        return headers
    }

    createMinimalOrder(data) {

        var url = this.SELCOM_BASE_URL + "/v1/checkout/create-order-minimal";
        axios.post(url, data, { headers: this.getHeaders(data) }).then(res => {
            console.log(res);
            return res.status(200).send(res.data);
        }).catch(err => {
            console.log(err.response);
            return err.response;
        });
    }


    createOrder(data) {
        var url = this.SELCOM_BASE_URL + "/v1/checkout/create-order"
        axios.post(url, data, { headers: this.getHeaders(data) }).then(res => {
            console.log(res);
            return res.status(200).send(res.data);
        }).catch(err => {
            console.log(err.response);
            return err.response;
        });
    }
    cretePushOnly(data) {
        var url = this.SELCOM_BASE_URL + "/v1/checkout/wallet-payment"
        axios.post(url, data, { headers: this.getHeaders(data) }).then(res => {
            console.log(res);
            return res.status(200).send(res.data);
        }).catch(err => {
            console.log(err.response);
            return err.response;
        });
    }

    creteMinimalOrderPush(min_data, push_data) {
        var url = this.SELCOM_BASE_URL + "/v1/checkout/wallet-payment"
        this.createMinimalOrder(min_data)
        axios.post(url, data, { headers: this.getHeaders(push_data) }).then(res => {
            console.log(res);
            return res.status(200).send(res.data);
        }).catch(err => {
            console.log(err.response);
            return err.response;
        });
    }

    getOrderStatus(order_id) {
        var url = this.SELCOM_BASE_URL + "/v1/checkout/order-status?id=" + order_id
        axios.get(url, data, { headers: this.getHeaders({ 'id': order_id }) }).then(res => {
            console.log(res);
            return res.status(200).send(res.data);
        }).catch(err => {
            console.log(err.response);
            return err.response;
        });
    }

    cancelOrder(order_id) {
        var url = this.SELCOM_BASE_URL + "/v1/checkout/cancel-order?id=" + order_id
        axios.get(url, data, { headers: this.getHeaders({ 'id': order_id }) }).then(res => {
            console.log(res);
            return res.status(200).send(res.data);
        }).catch(err => {
            console.log(err.response);
            return err.response;
        });
    }

    getAllOrders(fromDate, toDate) {
        url = this.SELCOM_BASE_URL + "/v1/checkout/list-orders?fromdate=" + fromDate + "&todate=" + toDate
        axios.get(url, data, { headers: this.getHeaders({ "fromdate": fromDate, "todate": toDate }) }).then(res => {
            console.log(res);
            return res.status(200).send(res.data);
        }).catch(err => {
            console.log(err.response);
            return err.response;
        });
    }
}

module.exports = new SelcomPayment();

File: routes/payment.py
import hmac, base64, datetime, time, hashlib, requests, json


class SelcomPay:
    SELCOM_API_KEY = "KUKUMARKET-8Plj0rPnRhVUkUSa"
    SELCOM_API_SECRET = "52695852-8a21-4f13-a885-a32a1a152161"
    SELCOM_BASE_URL = "https://apigwtest.selcommobile.com"

    utc_offset_sec = time.altzone if time.localtime().tm_isdst else time.timezone
    utc_offset = datetime.timedelta(seconds=-utc_offset_sec)
    timestamp = datetime.datetime.now().replace(tzinfo=datetime.timezone(offset=utc_offset)).isoformat()

    def __init__(self, *args, **kwargs):
        self.apiKeyEncoded = base64.b64encode(self.SELCOM_API_KEY.encode("utf-8"))

    def getDigest(self, data):
        message = 'timestamp=' + self.timestamp
        for d in data.keys():
            message = message + "&" + str(d) + "=" + str(data[d])
        return base64.b64encode(hmac.new(bytes(self.SELCOM_API_SECRET, "utf-8"), bytes(message, "utf-8"),
                                         digestmod=hashlib.sha256).digest()).decode()

    def getHeaders(self, data):
        headers = {
            "Content-type": "application/json",
            "Cache-Control": "no-cache",
            "Accept": "application/json",
            "Authorization": "SELCOM " + self.apiKeyEncoded.decode(),
            "Digest-Method": "HS256",
            "Digest": self.getDigest(data),
            "Timestamp": self.timestamp,
            "Signed-Fields": ",".join(data.keys())
        }
        return headers

    def createMinimalOrder(self, data):
        url = self.SELCOM_BASE_URL + "/v1/checkout/create-order-minimal"
        r = requests.post(url, data=json.dumps(data), headers=self.getHeaders(data), verify=True)
        return r.json()

    def createOrder(self, data, orderId):
        url = self.SELCOM_BASE_URL + "/v1/checkout/create-order"
        r = requests.post(url, data=json.dumps(data), headers=self.getHeaders(data), verify=True)
        return r.json()

    def cretePushOnly(self, orderId, data):
        url = self.SELCOM_BASE_URL + "/v1/checkout/wallet-payment"
        r = requests.post(url, data=json.dumps(data), headers=self.getHeaders(data), verify=True)
        return r.json()

    def creteMinimalOrderPush(self, min_data, push_data):
        url = self.SELCOM_BASE_URL + "/v1/checkout/wallet-payment"
        self.createMinimalOrder(min_data)
        time.sleep(4)
        r = requests.post(url, data=json.dumps(push_data), headers=self.getHeaders(push_data), verify=True)
        return r.json()

    def getOrderStatus(self, order_id):
        url = self.SELCOM_BASE_URL + "/v1/checkout/order-status?id=" + str(order_id)
        r = requests.get(url, headers=self.getHeaders({"id": order_id}))
        return r.json()

    def cancelOrder(self, order_id):
        url = self.SELCOM_BASE_URL + "/v1/checkout/cancel-order?id=" + str(order_id)
        r = requests.get(url, headers=self.getHeaders({"id": order_id}))
        return r.json()

    def getAllOrders(self, fromDate, toDate):
        url = self.SELCOM_BASE_URL + "/v1/checkout/list-orders?fromdate=" + str(fromDate) + "&todate=" + str(toDate)
        r = requests.get(url, headers=self.getHeaders({"fromdate": fromDate, "todate": toDate}))
        return r.json()
